classes/Module/AddonsCurlClient.php

```diff
@@ -0,0 +1,159 @@
+    {
+        $requestContent = $this->performRequest($type, $params);
+
+        return $this->parseRequest($requestContent);
+    }
+
+    /**
+     * @param string $type
+     * @param array $params
+     *
+     * @return string
+     */
+    private function performRequest($type, array $params)
+    {
+        if (!$this->isAddonsUp) {
+            return '';
+        }
+
+        $post_query_data = array(
+            'version' => isset($params['version']) ? $params['version'] : _PS_VERSION_,
+            'iso_code' => Tools14::strtolower(isset($params['iso_country']) ? $params['iso_country'] : 'all'),
+            'format' => isset($params['format']) ? $params['format'] : 'xml',
+        );
+        if (isset($params['source'])) {
+            $post_query_data['source'] = $params['source'];
+        }
+
+        $post_data = http_build_query($post_query_data);
+
+        $end_point = 'api.addons.prestashop.com';
+
+        switch ($type) {
+            case 'native':
+                $post_data .= '&method=listing&action=native';
+
+                break;
+
+            case 'service':
+                $post_data .= '&method=listing&action=service';
+
+                break;
+
+                break;
+            case 'must-have':
+                $post_data .= '&method=listing&action=must-have';
+
+                break;
+
+            case 'module':
+                $post_data .= '&method=module&id_module=' . urlencode($params['id_module']);
+                if (isset($params['username_addons'], $params['password_addons'])) {
+                    $post_data .= '&username=' . urlencode($params['username_addons']) . '&password=' . urlencode($params['password_addons']);
+                }
+
+                break;
+            case 'hosted_module':
+                $post_data .= '&method=module&id_module=' . urlencode((int) $params['id_module']) . '&username=' . urlencode($params['hosted_email'])
+                    . '&password=' . urlencode($params['password_addons'])
+                    . '&shop_url=' . urlencode(isset($params['shop_url']) ? $params['shop_url'] : Tools14::getShopDomain())
+                    . '&mail=' . urlencode(isset($params['email']) ? $params['email'] : Configuration::get('PS_SHOP_EMAIL'));
+
+                break;
+            case 'install-modules':
+                $post_data .= '&method=listing&action=install-modules';
+                $post_data .= defined('_PS_HOST_MODE_') ? '-od' : '';
+
+                break;
+            default:
+                return '';
+        }
+
+        $context = stream_context_create(array(
+            'http' => array(
+                'method' => 'POST',
+                'content' => $post_data,
+                'header' => 'Content-type: application/x-www-form-urlencoded',
+                'timeout' => 5,
+            ),
+        ));
+
+        if ($content = Tools14::file_get_contents('https://' . $end_point, false, $context)) {
```

Comment:
Only http and https change here, or I miss something? :thinking:

Reply:
Yes, it seems to be a fallback (I copied the code stupidly). I can use a loop like in `ps_mbo` if you prefer:

```php
        foreach ($protocols as $protocol) {
            if ($content = Tools::file_get_contents($protocol.'://'.$end_point, false, $context)) {
                return $content;
            }
        }
```